### PR TITLE
[Feature] Update legal hold status from generic message hint

### PR DESCRIPTION
**Pull request by johnxnguyen**
## What's new in this PR?

There are several ways to discover legal hold clients in a conversation, such as when backend reports missing clients when trying to send a message, or through an explicit sync of users clients. However, it could be that the user is acts as an observer in the conversation and rarely posts a message, if at all. 

To help the self client discover other legal hold clIents in the conversation, all messages received will contain a flag to indicate whether the sender believes their message was sent to a legal hold client. The self client uses this as a hint, first updating their local legal hold status for the conversation and subsequently triggering a sync to verify the legal hold status.

Verification relies on getting a complete list of all the participants in the conversation (using the empty message method) then syncing all unknown and all clients. It's important that while the client is doing this verification that it does not update the legal hold status triggered from other events because we can't be sure that we have enough information to make an accurate calculation of the legal hold state.

### Testing

Added unit tests that:

- the conversation is updated and the sync is triggered when needed
- generic message events are processed for the hints.
- the legal hold sync handle performs the sync an gives the result to the legal hold service

### Note

The `LegalHoldService` now takes on the responsibility of the `LegalHoldStatusUpdater` since it was getting a bit too messy to keep it separate and I don't see any good reason why they should be separate.

#### APK
[Download build #3449](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3449/artifact/build/artifact/wire-dev-PR3288-3449.apk)
[Download build #3475](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3475/artifact/build/artifact/wire-dev-PR3288-3475.apk)
[Download build #3496](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3496/artifact/build/artifact/wire-dev-PR3288-3496.apk)
[Download build #3497](http://10.10.124.11:8080/job/Pull%20Request%20Builder/3497/artifact/build/artifact/wire-dev-PR3288-3497.apk)

### CODE REVIEW

File: zmessaging/src/main/scala/com/waz/service/LegalHoldStatusUpdater.scala [hunk 1]
```diff
@@ -49,4 +57,32 @@ class LegalHoldStatusUpdaterImpl(clientsStorage: OtrClientsStorage,
...
+          case Some(updater) =>
+            convStorage.update(conv.id, updater)
+              .flatMap(_ => userService.syncClients(conv.id))
+          case None =>
+            Future.successful(())
+        }
+      case None =>
+        Future.successful(())
+    }
+  }
```
Comment (gizemb) [inline zmessaging/src/main/scala/com/waz/service/LegalHoldStatusUpdater.scala:73]: this looks a little hard to read. can we improve it by indenting second `match` further, or maybe using `fold`?

File: zmessaging/src/test/scala/com/waz/service/LegalHoldStatusUpdaterSpec.scala [hunk 2]
```diff
@@ -143,4 +143,51 @@ class LegalHoldStatusUpdaterSpec extends AndroidFreeSpec {
...
   }
 
+  feature("Generic message hints") {
+
+    scenario("it triggers client sync if status differ") {
+      // Given
+      val remoteConvId = RConvId("convId")
+      val convId = ConvId("convId")
+      val conv = ConversationData(convId, remoteConvId, legalHoldStatus = Disabled)
+      val messageStatus = Messages.LegalHoldStatus.ENABLED
```
Comment (gizemb) [inline zmessaging/src/test/scala/com/waz/service/LegalHoldStatusUpdaterSpec.scala:153]: would be nice to check the other case as well when status is Enabled, and message has Disabled

File: zmessaging/src/main/scala/com/waz/service/LegalHoldStatusUpdater.scala (hunk 1 above)
Comment (makingthematrix) [reply zmessaging/src/main/scala/com/waz/service/LegalHoldStatusUpdater.scala:73]: `for/yield` notation could also make it looks simpler.

**Review by makingthematrix — CHANGES_REQUESTED** [review]

File: zmessaging/src/test/scala/com/waz/service/LegalHoldStatusUpdaterSpec.scala (hunk 2 above)
Comment (johnxnguyen) [reply zmessaging/src/test/scala/com/waz/service/LegalHoldStatusUpdaterSpec.scala]: I added this in the refactoring.

File: zmessaging/src/main/scala/com/waz/service/LegalHoldStatusUpdater.scala (hunk 1 above)
Comment (johnxnguyen) [reply zmessaging/src/main/scala/com/waz/service/LegalHoldStatusUpdater.scala]: I went with the `for yield`.

**Comment by johnxnguyen** [comment]
@gizemb @makingthematrix I needed to do some major refactoring of the initial implementation, namely to post a client discovery message before doing the sync and to trigger the verification sync through the `SyncServiceHandle`.

**Review by makingthematrix — CHANGES_REQUESTED** [review]
Let's talk about it later, ok?
I believe this could be done using what we already have. But I need to know more.

**Review by makingthematrix — APPROVED** [review]

**Pull request merged**